### PR TITLE
8303764: Enable -Zc:wchar_t for Visual C++

### PR DESCRIPTION
**Pull request by TheShermanTanker**
Was: sunmscapi.dll cannot compile with Visual C++

`-Zc:wchar_t-` has, until now, been passed to switch off wchar_t as a distinct C++ type when compiling with Visual C++ so jchar and wchar_t are typedef'd to shorts on Windows. After some examination it appears this flag is not actually needed for almost every source file that we have, barring security.cpp, so we could try to remove it and make Windows code behave more like regular C++, since casting between wchar_t and jchar does not appear to happen very often (only appearing in one file!). wchar_t in Standard C is a typedef and is never affected by this flag, and if `-Zc:wchar_t` is enabled for C code, it is simply ignored

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303764](https://bugs.openjdk.org/browse/JDK-8303764): Enable -Zc:wchar_t for Visual C++

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12907/head:pull/12907` \
`$ git checkout pull/12907`

Update a local copy of the PR: \
`$ git checkout pull/12907` \
`$ git pull https://git.openjdk.org/jdk pull/12907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12907`

View PR using the GUI difftool: \
`$ git pr show -t 12907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12907.diff">https://git.openjdk.org/jdk/pull/12907.diff</a>

</details>

### CODE REVIEW

**Comment by TheShermanTanker** [comment]
@wangweij

**Comment by TheShermanTanker** [comment]
permissive- is a pretty big problem for Windows code in general, in particular our very jint and jlong typedefs themselves on Windows are completely broken on Windows and compiling with permissive- will straight up make the entire JDK explode. Change here is particular is a small part of a much bigger (JDK wide) fix, if that's what you meant by it happening in other files. I'm not sure if the code here could be redefined as a jchar from the start though, should I give it a try?

**Comment by djelinski** [comment]
Hi @TheShermanTanker What complier are you using? What command line?

`security.cpp` compiles just fine with `-permissive-` here, both debug and release. I'm using this compiler:
```
>cl.exe
Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31629 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```
you can find the cmdline under support/native/jdk.crypto.mscapi/libsunmscapi.

**Comment by TheShermanTanker** [comment]
> Hi @TheShermanTanker What complier are you using? What command line?
> 
> `security.cpp` compiles just fine with `-permissive-` here, both debug and release. I'm using this compiler:
> 
> ```
> >cl.exe
> Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31629 for x64
> Copyright (C) Microsoft Corporation.  All rights reserved.
> ```
> 
> you can find the cmdline under support/native/jdk.crypto.mscapi/libsunmscapi.

Strange...

```
C:\Program Files\Microsoft Visual Studio\2022\Community>cl.exe
Microsoft (R) C/C++ Optimizing Compiler Version 19.32.31332 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```

Seems like it's actually a version behind, too

**Comment by djelinski** [comment]
When `-Zc:wchar_t-` flag is added, `wchar_t` is equivalent to `jchar`; it's the default in C, which probably explains why this was the only failure. It allows us to avoid many reinterpret_casts, which are the root of all evil.

I guess we could explore changing jchar's typedef to wchar_t on Windows; we can't do that on other systems, because wchar_t is 32 bit there.

**Comment by TheShermanTanker** [comment]
Hmm, that sounds like an interesting prospect. char16_t might also be helpful here

**Comment by TheShermanTanker** [comment]
I'll experiment with switching off -Zc:wchar_t- in this change, will report back on any results

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp [hunk 1]
```diff
@@ -541,7 +541,7 @@ JNIEXPORT void JNICALL Java_sun_security_mscapi_CKeyStore_loadKeysOrCertificateC
                 else
                 {
                     if ((dwKeySpec & CERT_NCRYPT_KEY_SPEC) == CERT_NCRYPT_KEY_SPEC) {
-                        PP("CNG %I64d", (__int64)hCryptProv);
+                        PP("CNG %lld", (long long) hCryptProv);
```
Comment (djelinski) [inline src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:544]: What error were you seeing here before your changes?

Comment (TheShermanTanker) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:544]: Oh, that one wasn't really an error, but given that Microsoft has changed their definitions of a 64 bit int to a long long and the I64 specifier to ll in the ucrt for quite some time now, I figured that this was some extra cleanup I could fit in here. Not strictly required, but convenient code cleanliness for good measure nonetheless

**Comment by TheShermanTanker** [comment]
Oops, didn't mean to do that. Oh well, will update this again later

**Comment by magicus** [comment]
@TheShermanTanker This seems like a good fix to make VSC++ behave according to standard. I think it should be resurrected.

/open

**Comment by TheShermanTanker** [comment]
@magicus I think this was closed by GitHub itself since I messed up somewhere along the line with switching to a new branch to work on this. I could reopen a new Pull Request, but Daniel's concerns were what prevented me from doing so last time. Also, I think Skara may have a bug, since it seems to think that this closed Pull Request is open now

**Comment by magicus** [comment]
So, Daniel pointed out some irrelevant changes that you tried to sneak in. :) I think he is right that these should be reverted. I don't see anyone complaining about removing `-Zc:wchar_t-` though, or the necessary follow-up fixes for this. 

But this PR has indeed gotten a bit messy, so it's maybe better to open a new and just bring over the relevant fixes.

**Comment by TheShermanTanker** [comment]
> So, Daniel pointed out some irrelevant changes that you tried to sneak in. :) I think he is right that these should be reverted. I don't see anyone complaining about removing `-Zc:wchar_t-` though, or the necessary follow-up fixes for this.
> 
> But this PR has indeed gotten a bit messy, so it's maybe better to open a new and just bring over the relevant fixes.

I don't think Daniel's concerns stopped at the irrelevant changes though. I recall he was worried about removing the switch causing everywhere a jchar is passed to a WINAPI method needing a reinterpret cast, which is dangerous. I still don't know if that concern has been addressed yet

**Comment by djelinski** [comment]
If the changes in this PR are all that is required to enable the option, I think we can live with it.

**Pull request closed without merging**